### PR TITLE
[JENKINS-32225] - Add the functionality for Node usage tracking

### PR DESCRIPTION
**Pull request by oleg-nenashev**
- [x] - Add FlowListener extension point
- [x] - Add method allowing to retrieve executors from FlowExecutions
- [x] - Unit test demonstrating the use-case

http://issues.jenkins-ci.org/browse/JENKINS-32225

@reviewbybees

### CODE REVIEW

File: api/src/main/java/org/jenkinsci/plugins/workflow/flow/FlowListener.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    /**
+     * Notifies all listeners about the new flow node creation.
+     * @param node Newly created node
+     */
+    public static void fireNewHead(@Nonnull FlowNode node) {
+        for (FlowListener listener : all()) {
+            try {
+                listener.onNewHead(node);
+            } catch (Exception ex) { // Prevent failures on runtime exceptions
+                LOGGER.log(Level.SEVERE, "Runtime exception during the flow execution callback processing: " + listener, ex);
```
Comment (jglick) [inline api/src/main/java/org/jenkinsci/plugins/workflow/flow/FlowListener.java]: Copy `ItemListener.forAll` pending JENKINS-21224.

File: api/src/main/java/org/jenkinsci/plugins/workflow/flow/GraphListener.java [hunk 2]
```diff
@@ -27,6 +27,8 @@
 import org.jenkinsci.plugins.workflow.graph.FlowNode;
 
 /**
+ * A low-level implementation of the Flow listener.
+ * A classic Jenkins extension point is implemented as {@link FlowListener}.
```
Comment (jglick) [inline api/src/main/java/org/jenkinsci/plugins/workflow/flow/GraphListener.java]: Confusing. Why not just put the two new methods here in `GraphListener`?

File: api/src/main/java/org/jenkinsci/plugins/workflow/flow/FlowListener.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+public abstract class FlowListener implements ExtensionPoint {
+    
+    private final static Logger LOGGER = Logger.getLogger(FlowListener.class.getName());
+    
+    /**
+     * Notifies that a new execution node has been added to the flow.
+     * {@link FlowExecution} should batch up changes to a group and call this once, 
+     * as opposed to call this for every new node added.
+     *
+     * One of the use cases of this listener is to persist the state of {@link FlowExecution}.
```
Comment (jglick) [inline api/src/main/java/org/jenkinsci/plugins/workflow/flow/FlowListener.java]: Really? I do not think so. Looks like you just copied this from `GraphListener`.

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java [hunk 4]
```diff
@@ -66,11 +70,16 @@
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
+import static org.hamcrest.Matchers.*;
+import org.jenkinsci.plugins.workflow.support.steps.ExecutorStep;
+import org.jenkinsci.plugins.workflow.support.steps.ExecutorStepExecution;
```
Comment (jglick) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: Useless imports.

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java [hunk 5]
```diff
@@ -66,11 +70,16 @@
...
+import static org.junit.Assert.assertThat;
```
Comment (jglick) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: Can just switch to a wildcard static import (see lines 70–72).

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java [hunk 6]
```diff
@@ -414,5 +423,80 @@ private void startJnlpProc() throws Exception {
             }
         });
     }
+    
+    /**
+     * Ensures that {@link FlowListener} callback is being invoked for the {@link ExecutorStepExecution}.
+     * It allows to record the executor, which is being utilized by the flow.
+     */
+    @Test
```
Comment (jglick) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]:
```java
@Issue("JENKINS-32225")
```
?

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java [hunk 7]
```diff
@@ -414,5 +423,80 @@ private void startJnlpProc() throws Exception {
...
+        public void onNewHead(FlowNode node) { 
+            // We try to catch nodes, which indicate the start of the node() execution
+            if (!(node instanceof StepStartNode)) {
+                return;
+            }
+            
+            List<FlowNode> parents = node.getParents();
+            if (parents.isEmpty()) {
+                // Should never happen
+                return;
```
Comment (jglick) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]:
```java
assertNotEqual(Collections.emptyList(), parents);
```

File: support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java [hunk 8]
```diff
@@ -519,6 +520,49 @@ public String getUrl() {
             private static final long serialVersionUID = 1L;
         }
     }
+    
+    /**
+     * Looks for the executor, which is acquired by {@link ExecutorStep}.
+     * @param flowExecution Flow execution
+     * @return Found {@link Executor}. Null if it cannot be found
```
Comment (jglick) [inline support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java]: :bug: does not make sense. What is this supposed to return when there are two `node` blocks active in `parallel`?

I think you need to make it take `StepExecution`, or just be an instance method on `ExecutorStepExecution`.

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java [hunk 9]
```diff
@@ -414,5 +423,80 @@ private void startJnlpProc() throws Exception {
...
+     * It allows to record the executor, which is being utilized by the flow.
+     */
+    @Test
+    public void shouldInvokeFlowListenerCallbacksAndRecordNodes() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                DumbSlave s = createSlave(story.j);
+                s.setLabelString("remote quick");
+                s.getNodeProperties().add(new EnvironmentVariablesNodeProperty(
+                        new EnvironmentVariablesNodeProperty.Entry("ONSLAVE", "true")));
```
Comment (jglick) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: How is this variable relevant to the test?

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java [hunk 10]
```diff
@@ -414,5 +423,80 @@ private void startJnlpProc() throws Exception {
...
+                    "node('" + s.getNodeName() + "') {\n" +
+                    "    sh('echo before=`basename $PWD`')\n" +
+                    "    sh('echo ONSLAVE=$ONSLAVE')\n" +
+                    "}"));
 
+                // Run build and wait for its completion
+                startBuilding();
+                while (!e.isComplete()) {
+                    e.waitForSuspension();
+                }
```
Comment (jglick) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: Deprecated idiom, do not copy. Just

```java
story.j.assertBuildStatusSuccess(p.scheduleBuild2(0));
```

**Comment by jglick** [comment]
I am missing an overall description of what you are ultimately trying to accomplish here, and the JIRA issue does not give much background.

* The listener seems potentially useful, but
  * the `onStepExecutionStarted` and `onBodyExecutionStarted` events are both untested and unused
  * it might make more sense to use the existing `FlowExecution.addListener` for details, while adding a new API to notify global listeners of newly started or finished `FlowExecution`s (you can almost get there via `FlowExecutionOwner.Executable` + `RunListener` but you would need to poll on `getOrNull` unless `WorkflowRun.getExecutionPromise` were lifted into the interface)
* The method returning `Executor` strikes me as weird. And unnecessary: we already record `WorkspaceAction`, which should be all you need to determine from an `onNewHead` event that a node is being used. Without casting to `StepStartNode` or other tricks.

File: api/src/main/java/org/jenkinsci/plugins/workflow/flow/GraphListener.java (hunk 2 above)
Comment (oleg-nenashev) [inline api/src/main/java/org/jenkinsci/plugins/workflow/flow/GraphListener.java]: GraphListener is not a Jenkins extension point => it's external usage seems to be a hack.

File: api/src/main/java/org/jenkinsci/plugins/workflow/flow/FlowListener.java (hunk 3 above)
Comment (oleg-nenashev) [inline api/src/main/java/org/jenkinsci/plugins/workflow/flow/FlowListener.java]: This extension point may be used as well, but I agree it's a rudiment to be deleted from Javadoc

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java (hunk 5 above)
Comment (oleg-nenashev) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: Agreed

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java (hunk 6 above)
Comment (oleg-nenashev) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: yep

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java (hunk 7 above)
Comment (oleg-nenashev) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: agreed

File: support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java (hunk 8 above)
Comment (oleg-nenashev) [inline support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java]: Agreed. Maybe it was an over-engineering. I need to check why I did it

File: api/src/main/java/org/jenkinsci/plugins/workflow/flow/GraphListener.java (hunk 2 above)
Comment (jglick) [inline api/src/main/java/org/jenkinsci/plugins/workflow/flow/GraphListener.java]: Well you could make it implement `ExtensionPoint`.

Comment (oleg-nenashev) [inline api/src/main/java/org/jenkinsci/plugins/workflow/flow/GraphListener.java]: It's problematic due to the current plugin's API. It's hard to prevent usage conflicts since there are explicit methods for adding GraphListeners

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java (hunk 4 above)
Comment (oleg-nenashev) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: Nope. These classes belong to <code>org.jenkinsci.plugins.workflow.SUPPORT</code>, so they are not visible by default

File: support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java (hunk 8 above)
Comment (oleg-nenashev) [inline support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java]: Temporarily added some Javadoc

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java (hunk 9 above)
Comment (oleg-nenashev) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: yep, it's not required

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java (hunk 6 above)
Comment (oleg-nenashev) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: fixed

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java [hunk 11]
```diff
@@ -414,5 +420,77 @@ private void startJnlpProc() throws Exception {
...
+
+                // Run build and wait for its completion
+                story.j.assertBuildStatusSuccess(p.scheduleBuild2(0));
 
+                // Get the node invocation monitor
+                NodeInvocationListener listener = story.j.jenkins.getExtensionList(FlowListener.class)
+                        .get(NodeInvocationListener.class);
+                assertThat("NodeInvocationListener must be specified in the test", listener, notNullValue());
+                Node builtOn = listener.getBuiltOn();
+                assertThat("Node has not been recorded", builtOn, notNullValue());
```
Comment (amuniz) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: Using a negative sentence seems confusing, "Node has to be recorded" is better IMO so it reads "Assert that: node has to be recorded".

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java [hunk 12]
```diff
@@ -414,5 +420,77 @@ private void startJnlpProc() throws Exception {
...
+                assertThat("Node name differs from the expected one", builtOn.getNodeName(), equalTo(s.getNodeName()));
```
Comment (amuniz) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ExecutorStepTest.java]: Same here.

**Comment by amuniz** [comment]
:bee:

**Comment by jglick** [comment]
:bug: until the basic design is sorted out.

**Comment by jglick** [comment]
`FlowListener` is probably salvageable with some minor work, if there is a use case for it. `getExecutor` does not make sense as an API. Closing since this seems abandoned.

**Comment by oleg-nenashev** [comment]
@jglick Yes, the approach has been declined by the stakeholders. Currently we're working on another implementation
